airbnb.com: When trying to select days for checkin checkout, the date widget is empty

rniwa opened the pull request:
#### ecd009129377e8b137d771313e82be843dadaa59
<pre>
airbnb.com: When trying to select days for checkin checkout, the date widget is empty
<a href="https://bugs.webkit.org/show_bug.cgi?id=268058">https://bugs.webkit.org/show_bug.cgi?id=268058</a>

Reviewed by Chris Dumez and Wenson Hsieh.

This PR fixes a few bugs that caused idle callbacks to sometimes not get invoked.

1. Check for the existence of microtasks in shouldEndIdlePeriod and didReachTimeToRun were not taking
into account that there could be suspended microtasks (e.g. microtasks for a document in page cache).
Fixed this bug by introducing MicrotaskQueue::hasMicrotasksForFullyActiveDocument which checks for an
existence of a microtask for a fully active document.

2. When the event loop is sufficiently busy such that there is no idle period of at least 4ms,
opportunisticallyRunIdleCallbacks and didReachTimeToRun would perpetually defer idle callbacks.
Fixed this bug by decaying (shortening) the expected idle callback duration in each iteration using
exponential moving average with the weight of 50%. This will eventually shorten the expected idle
callback duration enough to start the idle period even when the event loop is constantly scheduling
timers and rAF callbacks.

3. opportunisticallyRunIdleCallbacks and didReachTimeToRun can sometimes exit without processing
pending idle callbacks and never schedule itself again when the initial check for idleness had failed.
Fixed this bug by always scheduling the event loop to run via scheduleToRunIfNeeded and in addition to
signaling OpportunisticTaskScheduler in IdleCallbackController::queueIdleCallback. This PR also moves
the code to signal OpportunisticTaskScheduler from Document to WindowEventLoop.

* LayoutTests/requestidlecallback/requestidlecallback-in-page-cache-with-pending-task.html: Added web
animation so that a microtask to stop the animation is queued as the document moves into the page cache.
* LayoutTests/requestidlecallback/requestidlecallback-is-called-even-when-event-loop-is-busy-expected.txt: Added.
* LayoutTests/requestidlecallback/requestidlecallback-is-called-even-when-event-loop-is-busy.html: Added.
* LayoutTests/requestidlecallback/requestidlecallback-is-called-when-idle-expected.txt: Added.
* LayoutTests/requestidlecallback/requestidlecallback-is-called-when-idle.html: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::protectedWindowEventLoop): Added.
(WebCore::Document::requestIdleCallback):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/IdleCallbackController.cpp:
(WebCore::IdleCallbackController::queueIdleCallback):
* Source/WebCore/dom/Microtasks.cpp:
(WebCore::MicrotaskQueue::hasMicrotasksForFullyActiveDocument const): Added.
* Source/WebCore/dom/Microtasks.h:
* Source/WebCore/dom/WindowEventLoop.cpp:
(WebCore::WindowEventLoop::scheduleIdlePeriod): Signal OpportunisticTaskScheduler and schedule the event
loop to run again.
(WebCore::WindowEventLoop::opportunisticallyRunIdleCallbacks):
(WebCore::WindowEventLoop::shouldEndIdlePeriod):
(WebCore::WindowEventLoop::didReachTimeToRun): Call shouldEndIdlePeriod instead of duplicating the code.
* Source/WebCore/dom/WindowEventLoop.h:
* Source/WebCore/page/OpportunisticTaskScheduler.h:

Canonical link: <a href="https://commits.webkit.org/276866@main">https://commits.webkit.org/276866@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f465b64911feecb6392635285b44cf6b169df5c7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45991 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25120 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48576 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48662 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42031 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48298 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29430 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22518 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37613 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46569 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22203 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39672 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18803 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19600 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40768 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4035 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42296 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41119 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50443 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20987 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17464 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44768 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22287 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/39830 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43660 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22646 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6406 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21981 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/IdleCallbackController.cpp:
```diff
@@ -60,7 +60,8 @@ int IdleCallbackController::queueIdleCallback(Ref<IdleRequestCallback>&& callbac
         });
-    }
+    } else if (RefPtr page = m_document ? m_document->page() : nullptr)
+        m_document->windowEventLoop().scheduleIdlePeriod(*page);
```
`windowEventLoop()` is not inline, we need a RefPtr to the document on the stack. Also, we should likely add `Document::protectedWindowEventLoop()` and call it.

rniwa replied:
Fixed.

whsieh commented on LayoutTests/requestidlecallback/requestidlecallback-is-called-when-idle.html:
```diff
@@ -0,0 +1,33 @@
+<html>
+<body>
+<p>This tests scheduling requestIdleCallback when the browser is idle. WebKit should fire idle callbacks.<br>
+You shouold see idle1, idle2, and idle3 below:</p>
```
Nit - typo

```suggestion
You should see idle1, idle2, and idle3 below:</p>
```

rniwa replied:
Fixed.

rniwa committed:
airbnb.com: When trying to select days for checkin checkout, the date widget is empty
https://bugs.webkit.org/show_bug.cgi?id=268058

Reviewed by Chris Dumez and Wenson Hsieh.

This PR fixes a few bugs that caused idle callbacks to sometimes not get invoked.

1. Check for the existence of microtasks in shouldEndIdlePeriod and didReachTimeToRun were not taking
into account that there could be suspended microtasks (e.g. microtasks for a document in page cache).
Fixed this bug by introducing MicrotaskQueue::hasMicrotasksForFullyActiveDocument which checks for an
existence of a microtask for a fully active document.

2. When the event loop is sufficiently busy such that there is no idle period of at least 4ms,
opportunisticallyRunIdleCallbacks and didReachTimeToRun would perpetually defer idle callbacks.
Fixed this bug by decaying (shortening) the expected idle callback duration in each iteration using
exponential moving average with the weight of 50%. This will eventually shorten the expected idle
callback duration enough to start the idle period even when the event loop is constantly scheduling
timers and rAF callbacks.

3. opportunisticallyRunIdleCallbacks and didReachTimeToRun can sometimes exit without processing
pending idle callbacks and never schedule itself again when the initial check for idleness had failed.
Fixed this bug by always scheduling the event loop to run via scheduleToRunIfNeeded and in addition to
signaling OpportunisticTaskScheduler in IdleCallbackController::queueIdleCallback. This PR also moves
the code to signal OpportunisticTaskScheduler from Document to WindowEventLoop.

* LayoutTests/requestidlecallback/requestidlecallback-in-page-cache-with-pending-task.html: Added web
animation so that a microtask to stop the animation is queued as the document moves into the page cache.
* LayoutTests/requestidlecallback/requestidlecallback-is-called-even-when-event-loop-is-busy-expected.txt: Added.
* LayoutTests/requestidlecallback/requestidlecallback-is-called-even-when-event-loop-is-busy.html: Added.
* LayoutTests/requestidlecallback/requestidlecallback-is-called-when-idle-expected.txt: Added.
* LayoutTests/requestidlecallback/requestidlecallback-is-called-when-idle.html: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::protectedWindowEventLoop): Added.
(WebCore::Document::requestIdleCallback):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/IdleCallbackController.cpp:
(WebCore::IdleCallbackController::queueIdleCallback):
* Source/WebCore/dom/Microtasks.cpp:
(WebCore::MicrotaskQueue::hasMicrotasksForFullyActiveDocument const): Added.
* Source/WebCore/dom/Microtasks.h:
* Source/WebCore/dom/WindowEventLoop.cpp:
(WebCore::WindowEventLoop::scheduleIdlePeriod): Signal OpportunisticTaskScheduler and schedule the event
loop to run again.
(WebCore::WindowEventLoop::opportunisticallyRunIdleCallbacks):
(WebCore::WindowEventLoop::shouldEndIdlePeriod):
(WebCore::WindowEventLoop::didReachTimeToRun): Call shouldEndIdlePeriod instead of duplicating the code.
* Source/WebCore/dom/WindowEventLoop.h:
* Source/WebCore/page/OpportunisticTaskScheduler.h:

Canonical link: https://commits.webkit.org/276866@main
